### ISSUE TITLE
SampleApp not connecting to AVS after restart.

### ISSUE THREAD

**Issue by vijay464** [alexa/avs-device-sdk#1319]
## Briefly summarize your issue:
I ran the sample and registered with AVS. 
For the first time it worked properly. After that I have restarted(ctrl+c and run again) the sample app. This time it is not connecting to AVS.
Attachec the debug log file. Please check.

## What is the expected behavior?
Whenever we run the sampleapp, It should connect to AVS and assist user if it is registered with AVS.

## What behavior are you observing?
SampleApp is not connecting to AVS

## Provide the steps to reproduce the issue, if applicable:
Run the sampleapp as per the document and register with AVS
Stop the sampleapp and run again.

## Tell us about your environment:
AVS Device SDK 1.12

### Tell us what hardware you're using:
- [ ] Desktop / Laptop

### Tell us about your OS (Type & version):
- [ ] Linux

[connectToAVSissue.txt](https://github.com/alexa/avs-device-sdk/files/3079813/connectToAVSissue.txt)

Thanks
Vijay

**Comment by visayamv** [comment]
Hi @vijay464 , in the log file you attached, did the logs stop there or are there more logs after 2019-04-15 10:44:19.430? I'd also like to know if there were any modifications done on the code?

**Comment by vijay464** [comment]
Hi visayamv,

I have done few modifications related to volume settings. But I did not change anything related to connecting to AVS. In the above attached file, I have collected only that logs . 
I faced this issue again and now I collected the more logs. Attached the new logfile here. Please check.


[AVSConnectionIssue.txt](https://github.com/alexa/avs-device-sdk/files/3116807/AVSConnectionIssue.txt)


Thanks
Vijay

**Comment by visayamv** [comment]
Hi Vijay, I'm not seeing the attempt to connect. Can you enable the sensitive logging: -DACSDK_EMIT_SENSITIVE_LOGS=ON to your CMAKE flags and post the logs here? Note that it may print out some credentials you send to AVS like the clientID, so make sure to scrub that before you post it here.

**Comment by kclchan** [comment]
I am closing this issue due to inactivity. Please feel free to re-open it if it has been closed in error.

**Comment by gabrieldespindula** [comment]
Hi folks, I think the issue I'm having was well described here, so I'm not creating a new one. Basically the sample app works fine when I first execute, after quitting by Ctrl + C and start over it doesn't work anymore, works again after some long interval (hours).
My sdk version is 1.20, using ubuntu 20. Following is the debug log after setting -DACSDK_EMIT_SENSITIVE_LOGS=ON.

[issueLogDbg9.txt](https://github.com/alexa/avs-device-sdk/files/5086805/issueLogDbg9.txt)

**Comment by mvelegon-amzn** [comment]
Hi @gabrieldespindula I don't see full logs here that would explain why the SDK is not able to connect to AVS after restart.

Can you attach full logs for 1) Fresh register and SDK is able to connect and 2) Restart

**Comment by gabrieldespindula** [comment]
Hi dear, today I wasn't able to make it work in the first try, however I did several attempts trying different options from the menu as well. Didn't succeed. The log is attached. I'll take a look in the logs myself and also keep trying a few more times, as soon as I manage to capture a successful run, I share it here.

[logSeveralTries.txt](https://github.com/alexa/avs-device-sdk/files/5098831/logSeveralTries.txt)

**Comment by gabrieldespindula** [comment]
Got a good log followed by a bad one after interrupting the execution using Ctrl + C.

[goodLog.txt](https://github.com/alexa/avs-device-sdk/files/5099722/goodLog.txt)

**Comment by mvelegon-amzn** [comment]
Hi @gabrieldespindula  there seem to be other issues with the logs you sent. For instance I see the following error show up. Note this is an issue with the audio on the device and should be fixed before using the SDK. Once you fix this issue please let us know if you are still seeing the connection error as you described earlier. 

```
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.rear
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.center_lfe
ALSA lib pcm.c:2642:(snd_pcm_open_noupdate) Unknown PCM cards.pcm.side
ALSA lib pcm_route.c:869:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:869:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:869:(find_matching_chmap) Found no matching channel map
ALSA lib pcm_route.c:869:(find_matching_chmap) Found no matching channel map
2020-08-19 22:46:19.773 [  2] 9 CurlEasyHandleWrapper:libcurl:id=1,text=successfully set certificate verify locations\:
2020-08-19 22:46:19.773 [  2] 9 CurlEasyHandleWrapper:libcurl:id=1,text=  CAfile\: /etc/ssl/certs/ca-certificates.crt
  CApath\: /etc/ssl/certs
2020-08-19 22:46:19.773 [  2] 9 CurlEasyHandleWrapper:libcurl:id=1,text=TLSv1.3 (OUT)\, TLS handshake\, Client hello (1)\:
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_oss.c:377:(_snd_pcm_oss_open) Unknown field port
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
ALSA lib pcm_usb_stream.c:486:(_snd_pcm_usb_stream_open) Invalid type for card
```

**Comment by gabrieldespindula** [comment]
Hi dear,
I was looking the logs and noticed both tries in the log (the good and the bad run) have this warnings regarding audio...

However today I was debugging and I found the issue.
Basically m_mutex is already locked when onAuthStateChange and/or sendMessage from HTTP2Transport are called. So it hangs when calling lock_guard. (I found simply swapping lock_guard to try_lock then trowing an error log when its already locked)
During the HTTP2Transport::mainLoop, it calls handleAuthorizing that locks m_mutex at  monitorSharedQueueWhileWaiting(State::AUTHORIZING) and someone else is calling HTTP2Transport::sendMessage after that, when m_mutex is already locked.

I made my own fix and it works here, but perhaps you guys know better the code and could bring a better solution.

Cheers
